src/app/complex/complex-results/complex-navigator/table-structure/table-interactor-column/table-interactor-column.component.ts

```diff
@@ -204,59 +209,105 @@ export class TableInteractorColumnComponent implements OnChanges {
+              if (enrichedComplex.startInteractorIndex === i) {
+                // The line starts in a subcomponent of the interactor, but not on the interactor itself,
+                // so the line does not start in the interactor when expanded
+                enrichedComplex.startInteractorIncludedWhenExpanded = false;
               }
+              enrichedComplex.endInteractorIndex = this.getMaxValue(enrichedComplex.endInteractorIndex, i);
+              // The subcomponent of this interactor is part of the complex, we update the start and end indices for the subcomponents
+              // line as it may start in this subcomponent
+              enrichedComplex.startSubComponentIndex = this.getMinValue(enrichedComplex.startSubComponentIndex, k);
+              enrichedComplex.endSubComponentIndex = this.getMaxValue(enrichedComplex.endSubComponentIndex, k);
             }
+          }
         }
       }
     }
     return enrichedComplex;
   }
 
   public classifyInteractorsByOrganism() {
-    this.enrichedInteractors.sort((a, b) => b.interactor.organismName.localeCompare(a.interactor.organismName));
+    this.enrichedInteractors.sort((a, b) => {
+      if (b.interactor.organismName === a.interactor.organismName) {
+        return b.timesAppearing - a.timesAppearing;
+      } else {
+        const organismBTimesAppearing = this._timesAppearingByOrganism.get(b.interactor.organismName);
+        const organismATimesAppearing = this._timesAppearingByOrganism.get(a.interactor.organismName);
+        if (organismBTimesAppearing === organismATimesAppearing) {
+          return b.interactor.organismName.localeCompare(a.interactor.organismName);
+        } else {
+          return organismBTimesAppearing - organismATimesAppearing;
+        }
+      }
+    });
     this.rangeOfInteractorOrganism();
   }
 
   public classifyInteractorsByType() {
-    this.enrichedInteractors.sort((a, b) => b.interactor.interactorType.localeCompare(a.interactor.interactorType));
+    this.enrichedInteractors.sort((a, b) => {
+      if (b.interactor.interactorType === a.interactor.interactorType) {
+        return b.timesAppearing - a.timesAppearing;
+      } else {
+        const typeBTimesAppearing = this._timesAppearingByType.get(b.interactor.interactorType);
+        const typeATimesAppearing = this._timesAppearingByType.get(a.interactor.interactorType);
+        if (typeBTimesAppearing === typeATimesAppearing) {
+          return b.interactor.interactorType.localeCompare(a.interactor.interactorType);
+        } else {
+          return typeBTimesAppearing - typeATimesAppearing;
+        }
+      }
+    });
     this.rangeOfInteractorType();
   }
 
   public classifyInteractorsByOccurrence() {
+    this.enrichedInteractors.sort((a, b) =>
+      b.timesAppearing - a.timesAppearing
+    );
+    this.ranges = [];
+  }
+
+  private calculateTimesAppearingType() {
```

Comment:
A couple of improvements on these methods. First, we could merge all 3 into a single method, to only iterate through interactors and components once. Second, we could simplify the has/get/set methods to deal with the map.

Regarding the maps
```
if (map.has(key)) {
    map.set(key, map.get(key) + 1);
} else {
    map.set(key, 1);
}
```
Is the same as
```
map.set((key, map.get(key) || 0) + 1);
```

Regarding merging the 3 methods (calculateTimesAppearingType, calculateTimesAppearingOrganism and interactorsTimesAppearing) into one, we could have
```
  private calculateTimesAppearing() {
    // Initialise times appearing by type or organism
    this._timesAppearingByType = new Map();
    this._timesAppearingByOrganism = new Map();
    for (const oneInteractor of this.enrichedInteractors) {
      // Initialise times appearing for each interactor
      oneInteractor.timesAppearing = 0;
      for (const complex of this.complexes) {
        const match = findInteractorInComplex(complex, oneInteractor.interactor.identifier, this.enrichedInteractors);
        if (!!match) {
          // Update times appearing for the interactor
          oneInteractor.timesAppearing += 1;
          // Update times appearing for the interactor type
          this._timesAppearingByType.set(
            oneInteractor.interactor.interactorType,
            (this._timesAppearingByType.get(oneInteractor.interactor.interactorType) || 0) + 1);
          // Update times appearing for the interactor organism
          this._timesAppearingByOrganism.set(
            oneInteractor.interactor.organismName,
            (this._timesAppearingByOrganism.get(oneInteractor.interactor.organismName) || 0) + 1);
        }
      }
    }
  }
```